java/arcs/core/data/expression/TypeEvaluator.kt

```diff
@@ -0,0 +1,155 @@
+    return when (expr.op) {
+      is Expression.UnaryOp.Not -> InferredType.Primitive.BooleanType
+      is Expression.UnaryOp.Negate -> expr.expr.accept(this, ctx)
+    }
+  }
+
+  override fun <L, R, T> visit(
+    expr: Expression.BinaryExpression<L, R, T>,
+    ctx: Scope
+  ): InferredType {
+    return when (expr.op) {
+      // Boolean ops
+      Expression.BinaryOp.And, Expression.BinaryOp.Or,
+      Expression.BinaryOp.LessThan, Expression.BinaryOp.GreaterThan,
+      Expression.BinaryOp.LessThanOrEquals,
+      Expression.BinaryOp.GreaterThanOrEquals,
+      Expression.BinaryOp.Equals,
+      Expression.BinaryOp.NotEquals -> InferredType.Primitive.BooleanType
+
+      // Numeric ops
+      Expression.BinaryOp.Add,
+      Expression.BinaryOp.Subtract,
+      Expression.BinaryOp.Multiply,
+      Expression.BinaryOp.Divide -> widen(
+        expr.left.accept(this, ctx),
+        expr.right.accept(this, ctx)
+      )
+
+      // Special ops
+      Expression.BinaryOp.IfNull -> expr.left.accept(this, ctx).union(expr.right.accept(this, ctx))
+    }
+  }
+
+  override fun <T> visit(expr: Expression.FieldExpression<T>, ctx: Scope): InferredType =
+    (if (expr.qualifier == null) {
+      ctx
+    } else {
+      (expr.qualifier.accept(this, ctx) as? ScopeType)?.scope
+    }).apply {
+      @Suppress("SENSELESS_COMPARISON") if (this == null && !expr.nullSafe) {
+        throw IllegalArgumentException("Field '${expr.field}' not looked up on null scope")
+      }
+    }?.lookup(expr.field) as InferredType
+
+  override fun <E> visit(expr: Expression.QueryParameterExpression<E>, ctx: Scope): InferredType {
+    return parameterScope.lookup(expr.paramIdentifier) as? InferredType
+      ?: throw IllegalArgumentException(
+        "Unbound parameter '${expr.paramIdentifier}'"
+      )
+  }
+
+  override fun visit(expr: Expression.NumberLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.TextLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.BooleanLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.NullLiteralExpression, ctx: Scope) =
+    InferredType.Primitive.NullType
+
+  private fun seqToScopeOr(inferredType: InferredType?, ctx: Scope): Scope =
+    ((inferredType as? SeqType)?.type as? ScopeType)?.scope ?: ctx
+
+  private fun InferredType.asScope() = (this as SeqType).type as ScopeType
+
+  override fun visit(expr: Expression.FromExpression, ctx: Scope): InferredType {
+    val scope = seqToScopeOr(expr.qualifier?.accept(this, ctx), ctx)
+
+    val resultSeq = expr.source.accept(this, scope) as SeqType
+    return SeqType(ScopeType(scope.builder().set(expr.iterationVar, resultSeq.type).build()))
+  }
+
+  override fun visit(expr: Expression.WhereExpression, ctx: Scope): InferredType {
+    return expr.qualifier.accept(this, ctx)
```

Comment:
as everywhere else, I would like us to verify types as well, Here I would expect us to evaluate the type of `expr.expr`(chicken chicken!) and ensure it is `Boolean`, wdyt?